src/lib/dataEntities/networks.ts

```diff
@@ -288,41 +288,65 @@ export const l2Networks: L2Networks = {
   },
 }
 
-const getNetwork = async (
-  signerOrProviderOrChainID: SignerOrProvider | number,
-  layer: 1 | 2
+const getNetworkByChainID = <TNetwork extends Network>(
+  chainID: number,
+  networks: NetworkList<TNetwork>
 ) => {
-  const chainID = await (async () => {
-    if (typeof signerOrProviderOrChainID === 'number') {
-      return signerOrProviderOrChainID
-    }
-    const provider = SignerProviderUtils.getProviderOrThrow(
-      signerOrProviderOrChainID
-    )
-
-    const { chainId } = await provider.getNetwork()
-    return chainId
-  })()
-
-  const networks = layer === 1 ? l1Networks : l2Networks
   if (networks[chainID]) {
     return networks[chainID]
-  } else {
-    throw new ArbSdkError(`Unrecognized network ${chainID}.`)
   }
+  throw new ArbSdkError(`Unrecognized network ${chainID}.`)
 }
 
-export const getL1Network = (
-  signerOrProviderOrChainID: SignerOrProvider | number
-): Promise<L1Network> => {
-  return getNetwork(signerOrProviderOrChainID, 1) as Promise<L1Network>
+const getChainIDfromSignerOrProvider = async (
+  signerOrProvider: SignerOrProvider
+) => {
+  const provider = SignerProviderUtils.getProviderOrThrow(signerOrProvider)
+  const { chainId } = await provider.getNetwork()
+  return chainId
 }
-export const getL2Network = (
-  signerOrProviderOrChainID: SignerOrProvider | number
-): Promise<L2Network> => {
-  return getNetwork(signerOrProviderOrChainID, 2) as Promise<L2Network>
+
+type NetworkOrPromiseOfNetwork<
+  TNetwork extends Network,
+  TSignerOrProviderOrNumber extends SignerOrProvider | number
+> = TSignerOrProviderOrNumber extends SignerOrProvider
+  ? Promise<TNetwork>
+  : TNetwork
+
+function getNetwork<
+  TNetwork extends Network,
+  TSignerOrProviderOrNumber extends SignerOrProvider | number
+>(
+  signerOrProviderOrChainID: TSignerOrProviderOrNumber,
+  networks: NetworkList<TNetwork>
+): NetworkOrPromiseOfNetwork<TNetwork, TSignerOrProviderOrNumber>
+function getNetwork<
+  TNetwork extends Network,
+  TSignerOrProviderOrNumber extends SignerOrProvider | number
+>(
+  signerOrProviderOrChainID: TSignerOrProviderOrNumber,
+  networks: NetworkList<TNetwork>
+): Promise<TNetwork> | TNetwork {
+  if (typeof signerOrProviderOrChainID === 'number') {
+    return getNetworkByChainID(signerOrProviderOrChainID, networks)
+  }
+  return getChainIDfromSignerOrProvider(signerOrProviderOrChainID).then(
+    chainID => getNetworkByChainID(chainID, networks)
+  )
 }
 
+export const getL1Network = <
+  TSignerOrProviderOrNumber extends SignerOrProvider | number
+>(
+  signerOrProviderOrChainID: TSignerOrProviderOrNumber
+) => getNetwork(signerOrProviderOrChainID, l1Networks)
```

Comment:
technically should probably return await here too, but probably doesnt matter if a bit of this stack trace gets skipped.